File: day4/solution4.2.ts
import * as fs from "fs";

function countXMAS(grid: string[]): number {
    const rows = grid.length;
    const cols = grid[0].length;
    let totalCount = 0;

    // Helper to check if X-MAS pattern exists at center (r, c)
    function isXMAS(r: number, c: number): boolean {
        if (
            r - 1 >= 0 && r + 1 < rows &&
            c - 1 >= 0 && c + 1 < cols
        ) {
            // Extract diagonals
            const topLeft = grid[r - 1][c - 1];
            const bottomRight = grid[r + 1][c + 1];
            const topRight = grid[r - 1][c + 1];
            const bottomLeft = grid[r + 1][c - 1];

            // Check for MAS in both diagonals
            const diagonal1 = topLeft + grid[r][c] + bottomRight;
            const diagonal2 = topRight + grid[r][c] + bottomLeft;

            return (
                (diagonal1 === "MAS" || diagonal1 === "SAM") &&
                (diagonal2 === "MAS" || diagonal2 === "SAM")
            );
        }
        return false;
    }

    // Traverse the grid
    for (let r = 1; r < rows - 1; r++) {
        for (let c = 1; c < cols - 1; c++) {
            if (grid[r][c] === "A" && isXMAS(r, c)) {
                totalCount++;
            }
        }
    }

    return totalCount;
}

// Read the grid from the input file
const grid = fs.readFileSync("input.txt", "utf-8").trim().split("\n");

const result = countXMAS(grid);
console.log(`Total occurrences of X-MAS: ${result}`);

File: day4/solution4.2.py
def count_x_mas(grid):
    rows = len(grid)
    cols = len(grid[0])
    total_count = 0

    # Helper function to check X-MAS pattern at center (r, c)
    def is_x_mas(r, c):
        if (
            0 <= r - 1 < rows and 0 <= r + 1 < rows and
            0 <= c - 1 < cols and 0 <= c + 1 < cols
        ):
            # Extract diagonals
            top_left = grid[r - 1][c - 1]
            bottom_right = grid[r + 1][c + 1]
            top_right = grid[r - 1][c + 1]
            bottom_left = grid[r + 1][c - 1]

            # Check for MAS in both diagonals
            diagonal_1 = top_left + grid[r][c] + bottom_right
            diagonal_2 = top_right + grid[r][c] + bottom_left

            return (diagonal_1 in ("MAS", "SAM") and diagonal_2 in ("MAS", "SAM"))
        return False

    # Traverse the grid
    for r in range(1, rows - 1):
        for c in range(1, cols - 1):
            if grid[r][c] == "A" and is_x_mas(r, c):
                total_count += 1

    return total_count


# Read the grid from the input file
with open("input.txt", "r") as f:
    grid = [line.strip() for line in f.readlines()]

result = count_x_mas(grid)
print(f"Total occurrences of X-MAS: {result}")